packages/fetchai/protocols/oef_search/message.py

```diff
@@ -43,6 +44,7 @@ class OefSearchMessage(Message):
     """A protocol for interacting with an OEF search service."""
 
     protocol_id = ProtocolId("fetchai", "oef_search", "0.1.0")
+    serializer = OefSearchSerializer
```

Comment:
From here I can see the issue:

`oef_search.serializer` requires the definition of `OefSearchMessage`  (it was already there)
`oef_search.message` requires the definition of `OefSearchSerializer` (added in this PR)

Some ideas:
- move the import of `OefSearchMessage` into the `OefSearchSerializer.encode|decode`. 
- populate the `OefSearchMessage.serializer` dynamically (at decode time is easy, but if the message is not from a decoding?)
- retrieve the serializer from the protocol registry via public id. I think what you are trying to achieve is that the connection knows how to serialize the message, given only the message. The public id is available, it just has to load the serializer associated with that public id.
- ... do we need a serializer? `encode/decode` could be abstract methods of `Message`... It is against the single responsibility principle, but in theory that could solve the problem. It's just not doable easily...

Reply:
Note, it has been solved above in the default protocol. The other protocols have not yet been finished. Re your suggestions:
-import in serializer code: not a fan
- dynamic not a fan either
- the problem with relying on public id is that we then need to maintain a protocol registry in the multiplexer (which for reasons highlighted above is potentially not ideal)
- yes i think we should maintain the separation into two classes.

The solution is actually surprisingly simple since we are dealing with a class variable. So a single assignment in the code base will satisfy our requirements. I now do this assignment in init:
``` python
from aea.protocols.default.message import DefaultMessage
from aea.protocols.default.serialization import DefaultSerializer

DefaultMessage.serializer = DefaultSerializer
```

Reply:
I don't think this it is a good solution. The user has to take into account yet another convention. If he forgets to update the serializer attribute, some code of some remote connection might fail quite unpredictably. This is another drop to the tons of other things users and developers have to take into account worsening user-experience and hence traction.

Reply:
This is done by the generator. The user does not need to do anything here. I think you need to look at how protocols are used in skills. The previous implementation was were convoluted to say the least; now we save a substantial amount of LOC! I think requiring more caution from the protocol designer (although it is auto generated, remember) than from the protocol user is preferable. The new approach makes using protocols MUCH simpler and it only adds one small requirement for the designer (which the generator takes care of).